### PR TITLE
Introduce earthly for local and CI builds on Github Actions

### PR DESCRIPTION
**Pull request by bjoernmeier**
Hi,
with this PR I want to switch to Github Actions as CI as travis is going to be closed.
As it was hard to reproduce build locally I want to introduce earthly as build tool to run the builds and tests on the CI the same way as run them locally. It did really help me.
What currently is missing and I would add later are tests on Windows and OS X. I hope we could start with the current matrix to enable builds again.
What I would drop is tests for pyarrow 0.16.x.

So please let me know what you think and what is missing or should be changed to bring this in.

### CODE REVIEW

**Comment by bjoernmeier** [comment]
Not sure how the settings of the repo are at the moment regarding enabling actions on branches. So this has to check an admin but you can see the lastest actions run on a PR in my fork https://github.com/bjoernmeier/turbodbc/actions/runs/541222301.

**Comment by xhochy** [comment]
@bjoern-meier-by Will review by EOD Tuesday, would be nice to have Github Actions activated by Wednesday so that we can merge this. Review will probably take a bit longer as I don't have `earthy` experience yet but I'm happy to learn about it.

File: docs/pages/contributing.rst [hunk 1]
```diff
@@ -49,6 +49,48 @@ current state.
...
+
+You can build turbodbc and run the full test suite with `earthly <https://earthly.dev>`_.
+
+See the install instructions on how to get earthly: `https://earthly.dev/get-earthly <https://earthly.dev/get-earthly>`_.
+
+Build turbodbc with the default arguments:
+
+::
+
+    earthly --allow-privileged +build
```
Comment (jochen-ott-by) [inline docs/pages/contributing.rst:63]: What is the "net result" of this build?
I assume this includes a python package (wheel and/or sdist?)? Where will it be (I could not find it locally, although from reading over the earth files, I though it should be somewhere on the host).

File: docs/pages/contributing.rst [hunk 2]
```diff
@@ -49,6 +49,48 @@ current state.
...
+
+::
+
+    earthly --allow-privileged +test-all
+
+Build and save a Docker image for development usage:
+
+::
+
+    earthly +docker
```
Comment (jochen-ott-by) [inline docs/pages/contributing.rst:87]: Some additional context would be nice here, e.g. how to run this image in docker? Also, it seems this earthly rule already runs the build, which (depending on what you're developing) might not be what you want ...

File: Earthfile [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        +test/result /result
+
+    SAVE ARTIFACT /result AS LOCAL result/$EARTHLY_TARGET_NAME
+
+test-python3.8-all:
+    BUILD test-python3.8-arrow0.x.x
+    BUILD test-python3.8-arrow1.x.x
+    BUILD test-python3.8-arrow2.x.x
+    BUILD test-python3.8-arrow3.x.x
+    BUILD test-python3.8-arrow-nightly:
```
Comment (jochen-ott-by) [inline Earthfile:225]: The last colon on this line seem off.

**Review by jochen-ott-by — CHANGES_REQUESTED** [review]

**Comment by MathMagique** [comment]
> Not sure how the settings of the repo are at the moment regarding enabling actions on branches. So this has to check an admin but you can see the lastest actions run on a PR in my fork https://github.com/bjoernmeier/turbodbc/actions/runs/541222301.

All actions are (and have been) enabled for this repo.

File: earthly/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    image: postgres:11
+    hostname: postgres
+    ports:
+      - 5432:5432
+    environment:
+      - POSTGRES_PASSWORD=password
+      - POSTGRES_USER=postgres
+      - POSTGRES_DB=test_db
+  mysql:
+    # test do not run with a newer mysql at the moment
```
[the remark this answers is not in the document]
Comment (bjoern-meier-by) [reply earthly/docker-compose.yml:15]: Yes, that was the idea that issues like that will go into issue tickets, did not have the time to report the issues for such problems. Will do this in the next days.

File: Earthfile [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        coveralls \
+        gmock \
+        gtest \
+        gxx_linux-64 \
+        mock \
+        pytest \
+        pytest-cov \
+        python=${PYTHON_VERSION} \
+        unixodbc \
+        boost-cpp \
```
[the remark this answers is not in the document]
Comment (xhochy) [reply Earthfile:72]: We should either use Ubuntu or `conda` as a source for the requirements. As both are full distributions, we should not mix them.

File: Earthfile [hunk 6]
```diff
@@ -0,0 +1,247 @@
+os-base:
+    ARG CODE_NAME=focal
+    FROM ubuntu:$CODE_NAME
```
Comment (xhochy) [inline Earthfile:3]: Use the docker image with Ubuntu and `conda`/`mamba` pre-installed, I'm also migrating them to focal (incl. the introduction of version tagged images) over at https://github.com/conda-forge/miniforge-images/pull/2 .
```suggestion
    FROM condaforge/mambaforge:latest
```

File: Earthfile [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    ENV MINICONDA_URL="https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh"
+    RUN cd /opt && \
+        wget --no-verbose -O miniconda.sh $MINICONDA_URL && chmod +x miniconda.sh
+    ENV MINICONDA=$HOME/miniconda
+    RUN cd /opt && \
+        ./miniconda.sh -b -p $MINICONDA && \
+        . $MINICONDA/etc/profile.d/conda.sh && \
+        conda config --remove channels defaults && \
+        conda config --add channels conda-forge && \
+        conda create -y -q -n turbodbc-dev \
```
Comment (xhochy) [inline Earthfile:58]: When using the `mambaforge` images, you don't need these steps anymore.

Also for the sake of feedback time, please create the environment with `mamba`. So much faster ;)
```suggestion
    RUN mamba create -y -q -n turbodbc-dev \
```

File: Earthfile [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        gcc_linux-64 \
```
Comment (xhochy) [inline Earthfile:59]: Use the platform independent metapackage `c-compiler`, this also helps when we want to run these tests on `aarch64`.
```suggestion
        c-compiler \
```

File: Earthfile [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        make \
+        ninja \
+        cmake \
+        coveralls \
+        gmock \
+        gtest \
+        gxx_linux-64 \
```
Comment (xhochy) [inline Earthfile:66]:
```suggestion
       cxx-compiler \
```

File: Earthfile [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        pytest \
+        pytest-cov \
+        python=${PYTHON_VERSION} \
+        unixodbc \
+        boost-cpp \
+        numpy$NUMPY_VERSION_RULE \
+        pyarrow$ARROW_VERSION_RULE \
+        pybind11 \
+        $CONDA_EXTRA \
+        -c conda-forge
```
Comment (xhochy) [inline Earthfile:77]: The channel is already given by the `conda` configuration.
```suggestion
        $CONDA_EXTRA
```

File: Earthfile [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+    RUN cd /opt && \
+        wget -q https://github.com/pybind/pybind11/archive/v2.6.2.tar.gz && \
```
[the remark this answers is not in the document]
Comment (xhochy) [reply Earthfile:80]: We should create a follow-up module to have `pybind11` as a build requirement in `pyproject.toml` and then use it either from `conda` in conda-based builds or let `pep517` pull it in automatically.

File: Earthfile [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+    FROM --build-arg CODE_NAME="$CODE_NAME" \
+        --build-arg PYTHON_VERSION="$PYTHON_VERSION" \
+        --build-arg ARROW_VERSION_RULE="$ARROW_VERSION_RULE" \
+        --build-arg NUMPY_VERSION_RULE="$NUMPY_VERSION_RULE" \
+        +build
+
+    RUN echo ". $MINICONDA/etc/profile.d/conda.sh" >> ~/.bashrc
+    RUN echo "conda activate turbodbc-dev" >> ~/.bashrc
+    RUN apt-get install -y vim
```
Comment (xhochy) [inline Earthfile:245]: 😻

**Comment by xhochy** [comment]
Close/reopened to trigger Github Actions.

**Comment by bjoern-meier-by** [comment]
@xhochy I think the issues is how fork PRs work in github. For PRs from forks the actions are invoked from the target main/master. For PRs changing the actions they have to be part of the repo. I will address the opened points from @jochen-ott-by here and then open a PR in the repo to see the actions work.
After they are on master I will check that Ps from branches really are run with the main's workflows.

File: Earthfile (hunk 11 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply Earthfile:80]: It is removed as a submodule in this PR, isn't it?

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+          - target: +test-python3.8-arrow3.x.x
+            experimental: false
+            upload: true
+          - target: +test-python3.8-arrow-nightly
+            experimental: true
+            upload: false
+    steps:
+      - uses: actions/checkout@v2
+      - name: Download latest earthly
+        run: "sudo /bin/sh -c 'wget https://github.com/earthly/earthly/releases/download/v0.5.0/earthly-linux-amd64 -O /usr/local/bin/earthly && chmod +x /usr/local/bin/earthly'"
```
[the remark this answers is not in the document]
Comment (bjoernmeier) [reply .github/workflows/ci.yml:35]: Did switch the naming to  5.0.0 as suggested

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Earthly version
+        run: earthly --version
```
[the remark this answers is not in the document]
Comment (bjoernmeier) [reply .github/workflows/ci.yml]: I did remove this as it is clear from above that version 5.0.0. is used. The bootstrap will only install thee bash and zsh completions which are not required here.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        if: ${{ matrix.upload }}
+        uses: codecov/codecov-action@v1
+        with:
+          directory: result/cov/cpp
+          fail_ci_if_error: true
+          verbose: true
+          functionalities: coveragepy, gcov
+
+      - name: Publish package
+        if: matrix.upload && env.IS_MASTER_BUILD == 'true' && github.event_name == 'push' && startsWith(github.ref, 'refs/tags') && github.repository == 'blue-yonder/turbodbc'
```
[the remark this answers is not in the document]
Comment (bjoernmeier) [reply .github/workflows/ci.yml:58]: Did add what you suggested. I have to be hones I could not check this line and we will probably see if it blocks uploads later when used.

File: Earthfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjoernmeier) [reply Earthfile:57]: Did remove duplicate apt-get installs as far as I could detect them as not necessary and used condo.

File: Earthfile (hunk 11 above)
[the remark this answers is not in the document]
Comment (bjoernmeier) [reply Earthfile:67]: I did forget to remove the submodule but did not use it anymore. Did remove it now as I think such a requirement for the builds is better suited inside the build env.
I agree with Uwe but would do the move pyproject.toml build requirements later and not in this PR.

File: docs/pages/contributing.rst [hunk 16]
```diff
@@ -49,6 +49,48 @@ current state.
...
+
+To run specific test targets and arguments, like different python or pyarrow versions, please see the Earthfile.
+
+Manual Setup
```
[the remark this answers is not in the document]
Comment (bjoernmeier) [reply docs/pages/contributing.rst]: Did update them to the state, that they are not a submodule anymore.

File: docs/pages/contributing.rst (hunk 1 above)
Comment (bjoernmeier) [reply docs/pages/contributing.rst]: Updated the earthly docs to represent this.

File: docs/pages/contributing.rst (hunk 2 above)
Comment (bjoernmeier) [reply docs/pages/contributing.rst:90]: Did update this section and added more information on who you can use this to run builds and tests.

File: Earthfile (hunk 3 above)
Comment (bjoernmeier) [reply Earthfile]: good catch, done

File: Earthfile (hunk 10 above)
Comment (bjoernmeier) [reply Earthfile]: Did switch to mambaforge and applied the above suggestions. Thanks, helped a lot.

**Comment by bjoernmeier** [comment]
will continue on https://github.com/blue-yonder/turbodbc/pull/292 as there the actions do run

**Pull request closed without merging**